src/v/redpanda/admin/security.cc

```diff
@@ -476,3 +496,60 @@ admin_server::oidc_revoke_handler(std::unique_ptr<ss::http::request>) {
     });
     co_return ss::json::json_return_type(ss::json::json_void());
 }
+
+ss::future<ss::json::json_return_type>
+admin_server::create_role_handler(std::unique_ptr<ss::http::request> req) {
+    if (need_redirect_to_leader(model::controller_ntp, _metadata_cache)) {
+        // In order that we can do a reliably ordered validation of
+        // the request (and drop no-op requests), run on controller leader;
+        throw co_await redirect_to_leader(*req, model::controller_ntp);
+    }
+
+    auto doc = co_await parse_json_body(req.get());
+
+    if (!doc.IsObject()) {
+        vlog(adminlog.debug, "Request body is not a JSON object");
+        throw_role_exception(
+          role_errc::malformed_def, "Request body is not a JSON object");
+    }
+
+    if (!doc.HasMember("role") || !doc["role"].IsString()) {
+        vlog(adminlog.debug, "String 'role' missing from request body");
+        throw_role_exception(
+          role_errc::malformed_def, "Missing string field 'role'");
+    }
+
+    auto role_name = security::role_name{doc["role"].GetString()};
+    validate_no_control(role_name(), string_conversion_exception{role_name()});
+
+    if (!security::validate_scram_username(role_name())) {
+        throw_role_exception(role_errc::invalid_name);
+    }
+
+    ss::httpd::security_json::role_definition j_res;
+    j_res.role = role_name();
+
+    if (is_no_op_role_write(
+          _controller->get_role_store().local(), role_name, security::role{})) {
+        vlog(adminlog.debug, "Empty role {} already exists", role_name);
+        co_return ss::json::json_return_type(j_res);
+    }
+
+    auto err
+      = co_await _controller->get_security_frontend().local().create_role(
+        role_name, security::role{}, model::timeout_clock::now() + 5s);
+
+    if (err == cluster::errc::role_exists) {
+        // Idempotency: if the empty role already exists,
+        // suppress the role_exists error and return success.
+        auto role = _controller->get_role_store().local().get(role_name);
```

Comment:
```suggestion
```

Reply:
This is now just duplicating the `auto role = ...` above.

Reply:
Right, I removed this now. I was thinking that it might be worth shortcutting a controller log entry replication by checking the existence of the role before calling `create_role`. But it does simplify the code a bit if we don't double-check.